### PR TITLE
feat(category) : add category CRUD feature

### PR DESCRIPTION
**Pull request by Yeseullyn**
Screenshot 2023-11-23 at 10 29 37 PM

Created category CRUD feature. I had a meeting with @hendrik-kim, backend category controller parameter will be updated, and the related agent and category slicer will also be updated accordingly.

### CODE REVIEW

File: frontend/src/components/pages/CategoryList.jsx [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+  // Use Redux hooks to access the redux store, dispatch action and retrieve data
+  const dispatch = useDispatch();
+  const categories = useSelector((state) => state.categories.categories);
+  const status = useSelector((state) => state.categories.status);
+  const error = useSelector((state) => state.categories.error);
+
+  // Set up local state using React useState for managing new category input and edited category data
+  const [newCategory, setNewCategory] = useState("");
+  const [editedCategory, setEditedCategory] = useState({ _id: "", name: "" });
```
Comment (hendrik-kim) [inline frontend/src/components/pages/CategoryList.jsx:21]: I believe that these two useState hooks could be improved by using only one.

File: frontend/src/components/pages/CategoryList.jsx [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+
+  // Dispatch the categoryListAsync when the component mounts to fetch the categoryList
+  useEffect(() => {
+    dispatch(categoryListAsync());
+  }, [dispatch]);
```
Comment (hendrik-kim) [inline frontend/src/components/pages/CategoryList.jsx:26]: Improve the category list to automatically update after adding or editing a category. However, when using useEffect, watch out for infinite loops: if the categoryListAsync action itself is updating the categories state, this code can cause an infinite loop

File: frontend/src/router/Routes.jsx [hunk 3]
```diff
@@ -24,6 +25,7 @@ export const router = createBrowserRouter([
       // { path: '/category', element: },
       // { path: 'blog-post', element: <BlogPostForm /> },
       { path: "/", element: <Home /> },
+      { path: "category", element: <CategoryList /> },
```
Comment (hendrik-kim) [inline frontend/src/router/Routes.jsx:28]: Please be a bit more specific about which route you are referring to.

File: frontend/src/slices/categorySlice.js [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+    } catch (error) {
+      return thunkAPI.rejectWithValue(error.message);
+    }
+  }
+);
+
+// Declare categorySlice with initial state and reducers
+const categorySlice = createSlice({
+  name: "categories",
+  initialState: {
```
Comment (hendrik-kim) [inline frontend/src/slices/categorySlice.js:63]: Improve your post-action processing logic to centralize it. For example, you could centralize the logic for fetching the category list back from the server after each successful action to add, delete, or update a category. Create a separate function for this and set it to call on each fulfilled state.

such as...

`// This function does the logic to fetch the list of categories back.
const fetchCategoriesAndUpdateState = async (dispatch) => {
    try {
        const updatedCategories = await dispatch(categoryListAsync()).unwrap();
        // This is where we can do any additional state updates or processing we need.
    } catch (error) {
        // Error handling logic
        console.error("Failed to fetch categories: ", error);
    }
}

////
Next, we call this function in the fulfilled state of each action:

const categorySlice = createSlice({
  name: "categories",
  // ... initialState, reducers, etc.

  extraReducers: (builder) => {
    builder
      .addCase(addCategoryAsync.fulfilled, async (state, action, { dispatch }) => {
        await fetchCategoriesAndUpdateState(dispatch);
      })
      .addCase(removeCategoryAsync.fulfilled, async (state, action, { dispatch }) => {
        await fetchCategoriesAndUpdateState(dispatch);
      })
      .addCase(updateCategoryAsync.fulfilled, async (state, action, { dispatch }) => {
        await fetchCategoriesAndUpdateState(dispatch);
      })
  },
});
`

**Review by hendrik-kim — APPROVED** [review]
Try to improve the code by incorporating the comments in your next task.

File: frontend/src/router/Routes.jsx (hunk 3 above)
Comment (Yeseullyn) [reply frontend/src/router/Routes.jsx:28]: Sure, thank!

File: frontend/src/components/pages/CategoryList.jsx (hunk 2 above)
Comment (Yeseullyn) [reply frontend/src/components/pages/CategoryList.jsx:26]: I omitted 'dispatch' from the dependency array to not cause an infinite loop. Thanks!

File: frontend/src/components/pages/CategoryList.jsx (hunk 1 above)
Comment (Yeseullyn) [reply frontend/src/components/pages/CategoryList.jsx:21]: I have modified code to make one useState hooks as your suggestion. Thanks!

**Pull request merged**